Add Bankless Leader design

aeyakovenko opened the pull request:
#### Problem

Leaders are a bottleneck on performance, since they have to do a lot of work of arranging transactions into well formed entries and shredding and broadcasting the entries to the rest of the network.  This CPU intensive work is competing for resources with the Bank.

#### Summary of Changes

Propose a design in which leaders skip execution of transactions and produce valid entries by only verifying that the transaction can afford to be included in the block.

Fixes #

mvines commented on book/src/proposals/bankless-leader.md:
```diff
@@ -0,0 +1,59 @@
+of the block the cache is destroyed.
+
+## Balance Check
+
```
What signatures are verified in the Transaction?  Just the execution key's signature or all?

aeyakovenko replied:
all are verified

mvines commented on book/src/terminology.md:
```diff
@@ -74,6 +74,10 @@ A globally unique identifier that is also a proof that the [entry](terminology.m
 
+## execution key
+
+The key in the transaction that pays for the cost of including the transaction in the ledger.  This is the first key in the transaction.
```
Worth suggesting here that an execution key is highly recommended to be set as Credit-Only?

aeyakovenko replied:
@mvines no, its CD to pay for the fee.  it just can't be used in any instruction.

aeyakovenko replied:
@mvines, I think it must be CD, otherwise the CO assumptions will fail.

mvines replied:
Yes my bad.

aeyakovenko commented on book/src/proposals/bankless-leader.md:
```diff
@@ -0,0 +1,59 @@
+
+While a validator only needs to reassemble the block and replay
+execution of well formed entries.  For transactions that have
+low execution costs, the leader does 3x more memory operations than
```
Both.  Just looking at how transactions flow through the leader, they come unsorted in random packets at the ingress point, then they need to be packed into entries of accepted transactions, then they need to be shredded.  replay stage doesn't need to move any memory besides landing the shreds into the block.

mvines commented:
Seems like we could back up a sunny day prototype of this and measure the TPS impact pretty easily

garious commented on book/src/proposals/bankless-leader.md:
```diff
@@ -0,0 +1,65 @@
+low execution costs, the leader does 3x more memory operations
+before any bank execution than the validator.
+
+## Execution Key
```
How about we call this a "Fee Account"?

garious commented:
Seems to me that this would be at most a 2x boost in the leader's Banking Stage, because it allows that stage to load one account where it would otherwise load two. And that assumes that loading accounts is the performance bottleneck where doing simple System transfers (like in bench-tps).  When doing BPF smart contracts (say bench-exchange), I'd expect this would be a much bigger optimization.   If that's the case, that the bench-exchange is significant and bench-tps is insignificant, we should consider relaxing this proposal a bit, and running all System transfers on the leader. By running all System transactions, the user could transfer tokens to other users without it impacting their ability to use that account for subsequent transaction fees.

aeyakovenko commented:
@garious leader needs to do 2 loads and 2 stores.  With this proposal leader just does 1 load.  so 5x less account work.  The stores might take a bit more time than reads as well.

rob-solana commented:
the leader isn't exempted from the work of banking in this proposal, the work is merely moved to replay

consecutive leader blocks suffer in this proposal, as ingest has to wait for a bank from replay

aeyakovenko commented:
@rob-solana Yea, that is true.  But, the current leader can catch up in the replay stage.  

bankless leader + replay stage 

Only does 1 more additional load, which is loading the fee account twice.  But the additional loads should be amortized, since they will remain in memory for the replay stage.

Replay stage is easier to optimize, all the transactions are already in entires.  If its my own leader's block, I can trust that they can all be executed in parallel without the additional check.

aeyakovenko commented:
> consecutive leader blocks suffer in this proposal, as ingest has to wait for a bank from replay
@rob-solana why is this true?  The next block can be encoded, PoH doesn't depend on bankhash, only the blockhash.

rob-solana commented:
> @rob-solana Yea, that is true. But, the current leader can catch up in the replay stage.
> 
> bankless leader + replay stage
> 
> Only does 1 more additional load, which is loading the fee account twice. But the additional loads should be amortized, since they will remain in memory for the replay stage.
> 
> Replay stage is easier to optimize, all the transactions are already in entires. If its my own leader's block, I can trust that they can all be executed in parallel without the additional check.

I suppose we can also skip PoH verification, but I still see banking_stage for a new block waiting until all the work of actually banking is finished.  That's ok because the rest of the network is getting their Entries sooner?

rob-solana commented:
> > consecutive leader blocks suffer in this proposal, as ingest has to wait for a bank from replay
> 
> @rob-solana why is this true? The next block can be encoded, PoH doesn't depend on bankhash, only the blockhash.

the next block depends on _balances_ from the previous block

garious commented:
Would it make sense to rephrase this proposal as "split BankingStage into FeeVerify+ReplayStage and then put Broadcast before ReplayStage"?  Then the proposal would just be about when and how ReplayStage needs to block FeeVerify for the following slot.

aeyakovenko commented:
> Would it make sense to rephrase this proposal as "split BankingStage into FeeVerify+ReplayStage and then put Broadcast before ReplayStage"? Then the proposal would just be about when and how ReplayStage needs to block FeeVerify for the following slot.

@garious it doesn't need to block it.  just the cache needs to be consistent with the parent its based on.  as soon as replay is done, the cache can be reset on the next block.

aeyakovenko commented:
@rob-solana I think I addressed your concerns.  "fee verify" is aware of the base fork its caching, and as soon as replay is done with a new bank it can reset its cache on the next block.  In theory it should lag by at most 2 blocks.

rob-solana commented on book/src/proposals/bankless-leader.md:
```diff
@@ -65,6 +66,19 @@ When the leader finishes the replay stage it can reset the balance
+block and create an empty cache.
+
+2. if the cache is initialized, check if block's parents has a new
+finalized bank that is newer than the current base fork for the
```
frozen bank?

aeyakovenko replied:
@rob-solana, yea correct.

rob-solana commented:
yeah, I think my concerns are addressed

sakridge commented:
> @garious
> 
> > split BankingStage into FeeVerify+ReplayStage
> 
> Replay is asynchronous to FeeVerify, so not sure if that makes sense. But I like the name FeeVerify as the stage. How about “Replace BankingStage with FeeVerify for the leader”

If we are doing FeeVerify *and* replay at the same time most of the time then seems like we are doing even more work than before.

garious commented:
> If we are doing FeeVerify _and_ replay at the same time most of the time then seems like we are doing even more work than before.

No, that's what Anatoly is saying about "asynchronous".  We just need to watch for places where FeeVerify can't proceed until ReplayStage executes the previous block.

aeyakovenko commented:
@sakridge two different pipelines which shouldn’t block each other if you have enough cores.

sakridge commented:
> @sakridge two different pipelines which shouldn’t block each other if you have enough cores.

If we had enough cores, we wouldn't need this design because broadcast wouldn't be impacted. I thought this whole design is to free up cores for broadcast, but now you are saying you need even more cores to run replay and fee verify and broadcast in parallel.

aeyakovenko commented:
@sakridge that’s not accurate.  If you have 1 pipeline of size N, vs 2  pipelines of size N/2, the latter can achieve better latency with more cores.
